### ISSUE TITLE
PR & Unit tests

### ISSUE THREAD

**Issue by michaelw85** [MicrosoftDocs/vsts-docs#3284]
If your flow would involve a pull request including unit tests that have to pass before merging and than publishing a package I would think it is possible to determine quality before releasing. Any build that fails would never result in a merge and would mean will never be released.
Would views still give any added value in such a workflow?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1a891e9b-bf26-ea09-1da3-ebe3e57b7e26
* Version Independent ID: bf53c51c-c1c1-1931-44ef-81433746e5ec
* Content: [Views on Azure DevOps Services feeds](https://docs.microsoft.com/en-us/azure/devops/artifacts/concepts/views?view=azure-devops#feedback)
* Content Source: [docs/artifacts/concepts/views.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/artifacts/concepts/views.md)
* Product: **devops**
* GitHub Login: @elbatk
* Microsoft Alias: **elbatk**

**Comment by seangwright** [comment]
@michaelw85 I had a similar thought.

TLDR: Views seem to be another tool beyond package Semver suffixes and both can be used together but might be overkill for your use-case.

My current release pipeline uses a build artifact that is the build output of `dotnet build`. A NuGet package is created during the stages of my release pipeline.

The first stage creates a NuGet package which applies a build suffix of `ci-<build-number>` to the package version.

This package is published to the `@local` view.

The next stage requires a manual approval of the previous stage (dev goes into DevOps and clicks "approve" on the release stage). This helps to affirm that the quality of the package is sufficient to move to the next view.

The 2nd stage promotes the previously created package to the `@beta` view.

To move past this stage requires the same manual approval as moving past stage 1 does.

The 3rd stage re-packages the build artifact without the `ci-<build_number>` suffix, publishes the package to the `@local` view and then promotes it to the `@release` view.

The `@release` view is considered to be the view that hosts packages which can be integrated for production use.

`@local` means the build artifact built and passed tests successfully but hasn't been tested for integration into a codebase.

`@beta` means the package has been tested being integrated but needs final sign-off before being approved for production use.

All that said... I could probably get away with using only the `@local` view and doing the re-package w/o the `ci-` suffix and publishing of the package as my 2nd stage. The first stage would be the creation of the `ci-` suffixed package pushed to `@local`.

**Comment by michaelw85** [comment]
@seangwright Thanks for sharing. I will keep this in mind, it's similar to something I was thinking about but it is overkill for me (us). It sounds like overhead that does not necessarily add to drastic quality improvements.

In my case our packages are for internal use only and the product consuming them will have it's own ci/cd DTAP release flow. I think gated check-ins would "guarantee" enough quality for us before consuming the package in development, any issues can be addressed quite quickly and the consuming product will not automatically update to latest anyway. The product will be build and than moved through the DTAP stages, nothing changes apart from config input so new package versions can continue to be released at all times without impacting existing releases of a product that consumes such a package.

**Comment by NightOwl888** [comment]
@seangwright

I like your setup, but I am curious how you would handle the scenario for multiple NuGet packages. For example, let's say ProjA depends on ProjB, which in turn depends on ProjC. They are all in separate repositories with dissimilar version numbers, but a new version of each is being developed simultaneously.

ProjA = 1.6.0-beta0003
ProjB = 1.4.0-beta0004
ProjC = 1.2.0-beta0001

So, clearly ProjC has to be deployed first, which puts it at version 1.2.0, so what sort of logic do you have to make sure that ProjB gets 1.4.0 as its version number and its package reference to ProjC gets updated to 1.2.0 from its beta counterpart (or wildcard)?

Also, are you using .nuspec files to repackage, or have you found a way to take advantage of the [.csproj PackageReferences](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj#packagereference) and [NuGet metadata properties](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj#nuget-metadata-properties) to repackage the files?

**Comment by Rutix** [comment]
@michaelw85 I think it depends on what the scope of validation is to you. For us the validation part is not only technical unittest but we also want to validate the consume to see if we didn't break anything functional before exposing it to the other consumers.

**Comment by chasewilson** [comment]
Hi all. Thank you for the feedback, questions, and input. I'm going to be closing this issue for now. There will be some updates to view usage and other related content. If you have any more questions or issues with the documentation, feel free to create a new issue!
